// @discordx/pagination/build/cjs/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PaginationType = exports.SelectMenuPageId = exports.defaultIds = exports.defaultTime = void 0;
// By default, five minute.
exports.defaultTime = 3e5;
const prefixId = "discordx@pagination@";
exports.defaultIds = {
    buttons: {
        end: prefixId + "endButton",
        exit: prefixId + "closeButton",
        next: prefixId + "nextButton",
        previous: prefixId + "previousButton",
        start: prefixId + "startButton",
    },
    menu: prefixId + "menu",
};
var SelectMenuPageId;
(function (SelectMenuPageId) {
    SelectMenuPageId[SelectMenuPageId["Start"] = -1] = "Start";
    SelectMenuPageId[SelectMenuPageId["End"] = -2] = "End";
    SelectMenuPageId[SelectMenuPageId["Exit"] = -3] = "Exit";
})(SelectMenuPageId = exports.SelectMenuPageId || (exports.SelectMenuPageId = {}));
var PaginationType;
(function (PaginationType) {
    PaginationType[PaginationType["Button"] = 0] = "Button";
    PaginationType[PaginationType["SelectMenu"] = 1] = "SelectMenu";
})(PaginationType = exports.PaginationType || (exports.PaginationType = {}));
